src/ccall.cpp

```diff
@@ -1648,166 +1728,242 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
-        bool issigned = jl_signed_type && jl_subtype(jargty, (jl_value_t*)jl_signed_type);
-        argvals[ai + sret] = llvm_type_rewrite(v, largty,
-                ai + sret < fargt_sig.size() ? fargt_sig.at(ai + sret) : fargt_vasig,
-                false, byRef, issigned, ctx);
+        if (isa<UndefValue>(v)) {
+            JL_GC_POP();
+            return jl_cgval_t();
+        }
+        assert(v->getType() == pargty);
+        argvals[ai + sret] = v;
     }
 
+    Value *result = NULL;
+    // First, if the ABI requires us to provide the space for the return
+    // argument, allocate the box and store that as the first argument type
+    bool sretboxed = false;
+    if (sret) {
+        jl_cgval_t sret_val = emit_new_struct(rt, 1, NULL, ctx); // TODO: is it valid to be creating an incomplete type this way?
+        assert(sret_val.typ != NULL && "Type was not concrete");
+        if (!sret_val.ispointer()) {
+            Value *mem = emit_static_alloca(lrt, ctx);
+            builder.CreateStore(sret_val.V, mem);
+            result = mem;
+        }
+        else {
+            // XXX: result needs a GC root here if result->getType() == T_pjlvalue
+            result = sret_val.V;
+        }
+        argvals[0] = emit_bitcast(result, fargt_sig.at(0));
+        sretboxed = sret_val.isboxed;
+    }
+
+    Instruction *stacksave = NULL;
+    if (needStackRestore) {
+        stacksave = CallInst::Create(Intrinsic::getDeclaration(jl_Module,
+                                                               Intrinsic::stacksave));
+        if (savespot) {
+#if JL_LLVM_VERSION >= 30800
+            instList.insertAfter(savespot->getIterator(), stacksave);
+#else
+            instList.insertAfter(savespot, stacksave);
+#endif
+        }
+        else {
+            instList.push_front(stacksave);
+        }
+    }
 
     // make LLVM function object for the target
     // keep this close to the function call, so that the compiler can
     // optimize the global pointer load in the common case
     Value *llvmf;
-    FunctionType *functype = FunctionType::get(sret ? T_void : prt, fargt_sig, isVa);
-
-    if (jl_ptr != NULL) {
-        null_pointer_check(jl_ptr,ctx);
-        Type *funcptype = PointerType::get(functype,0);
-        llvmf = builder.CreateIntToPtr(jl_ptr, funcptype);
+    if (llvmcall) {
+        if (symarg.jl_ptr != NULL) {
+            jl_error("llvmcall doesn't support dynamic pointers");
+        }
+        else if (symarg.fptr != NULL) {
+            jl_error("llvmcall doesn't support static pointers");
+        }
+        else if (symarg.f_lib != NULL) {
+            jl_error("llvmcall doesn't support dynamic libraries");
+        }
+        else {
+            assert(symarg.f_name != NULL);
+            llvmf = jl_Module->getOrInsertFunction(symarg.f_name, functype);
+            if (!isa<Function>(llvmf) || cast<Function>(llvmf)->getIntrinsicID() == Intrinsic::not_intrinsic)
```

Comment:
As mention by Tim in https://github.com/JuliaLang/julia/pull/18754#issuecomment-250977056 it would be nice if we could turn off the second check. Maybe a second calling convention `llvmcall_unsafe`?